src/v/raft/state_machine_base.h

```diff
@@ -97,6 +119,28 @@ class state_machine_base {
         return snapshot_at_offset_supported::yes;
     }
 
+    /**
+     * Returns true only if this is the first start of the state machine and it
+     * needs an initial recovery to happen. In this case the state machine
+     * manager will perform the initial recovery according to the returned
+     * policy.
+     *
+     * This method is called after the state machine has been started but before
+     * any apply happened.
+     */
+    virtual bool needs_initial_recovery() const {
+        return last_applied_offset() == model::offset{};
+    }
+
+    /**
+     * Returns state machine configured initial recovery policy. The default
+     * policy is to read everything as this is the basic behavior of the state
+     * machine
+     */
+    virtual stm_initial_recovery_policy get_initial_recovery_policy() const {
+        return stm_initial_recovery_policy::read_everything;
```

Comment:
Not sure about this being a good default, for new state machines this is probably should be `skip_to_end`, but discoverability is low - it is easy to forget to change it. Should this be a pure virtual instead to force the stm authors to make a choice? (could be a followup)